### PR TITLE
feat: add RPC to return median fee rate

### PR DESCRIPTION
**Pull request by chanhsu001**
add RPC to return median fee rate

### What problem does this PR solve?

Add new rpc interface to fetch transaction median fee rate
user could specify blocks_to_scan and min_transactions_in_block parameter to filter blocks or use default setting.

Problem Summary:

### What is changed and how it works?

when new block is added,  also insert BlockTxStat, contains all transactions (exclude cellbase tx) cycle and tx_size information; 
we add new RPC to support scan number of latest blocks on chain, collect number of transactions; and calculate the median fee rate

What's Changed:

### Related changes

- db-schema

Tests 

- Unit test

Side effects

- db-schema change

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: db-schema/src/lib.rs [hunk 1]
```diff
@@ -38,6 +38,8 @@ pub const COLUMN_NUMBER_HASH: Col = "13";
 pub const COLUMN_CELL_DATA_HASH: Col = "14";
 /// Column store block extension data
 pub const COLUMN_BLOCK_EXTENSION: Col = "15";
+/// Column store block's transaction_cycle_size
```
Comment (quake) [inline db-schema/src/lib.rs:41]:
```suggestion
/// Column store block's transactions cycle and size
```

File: db-schema/src/lib.rs [hunk 2]
```diff
@@ -38,6 +38,8 @@ pub const COLUMN_NUMBER_HASH: Col = "13";
...
+pub const COLUMN_TX_STAT: Col = "16";
```
Comment (quake) [inline db-schema/src/lib.rs:42]:
```suggestion
pub const COLUMN_BLOCK_TXS_STAT: Col = "16";
```

File: chain/src/chain.rs [hunk 3]
```diff
@@ -763,13 +763,34 @@ impl ChainService {
...
+                                    // block tx statistics will not include cellbase tx
+                                    // entry.cycles already exclude cellbase
+                                    let tx_stat = BlockTxStat {
+                                        txs_cycles: cache_entries
+                                            .into_iter()
+                                            .map(|entry| entry.cycles)
+                                            .collect(),
+                                        txs_size: resolved
+                                            .iter()
+                                            .filter(|tx| !tx.is_cellbase())
```
Comment (quake) [inline chain/src/chain.rs:783]: first tx of block is always cellbase, we may use skip:
```suggestion
                                            .skip(1)
```

File: chain/src/chain.rs [hunk 4]
```diff
@@ -763,13 +763,34 @@ impl ChainService {
...
                                     let mut mut_ext = ext.clone();
                                     mut_ext.verified = Some(true);
                                     mut_ext.txs_fees = txs_fees;
                                     txn.insert_block_ext(&b.header().hash(), &mut_ext)?;
+
+                                    // block tx statistics will not include cellbase tx
+                                    // entry.cycles already exclude cellbase
+                                    let tx_stat = BlockTxStat {
+                                        txs_cycles: cache_entries
+                                            .into_iter()
```
Comment (quake) [inline chain/src/chain.rs:778]:
```suggestion
                                            .iter()
```

File: rpc/README.md [hunk 5]
```diff
@@ -1372,6 +1373,51 @@ Response
...
+* `get_median_fee_rate(blocks_to_scan, num_transactions)`
+    * `blocks_to_scan`: [`Uint32`](#type-uint32) `|` `null`
+    * `num_transactions`: [`Uint32`](#type-uint32) `|` `null`
+* result: [`Capacity`](#type-capacity)
+
+Returns the median fee rate in recent number of blocks specified by user.
+
+##### Params
+
+*   `blocks_scan` - the latest x (default and maximum blocks: 42) blocks scanned to calculate median fee rate.
```
Comment (quake) [inline rpc/README.md:1386]: wrong params name

```suggestion
*   `blocks_to_scan` - the latest x (default and maximum blocks: 42) blocks scanned to calculate median fee rate.
```

File: rpc/README.md [hunk 6]
```diff
@@ -1372,6 +1373,51 @@ Response
 }
 ```
 
+#### Method `get_median_fee_rate`
+* `get_median_fee_rate(blocks_to_scan, num_transactions)`
+    * `blocks_to_scan`: [`Uint32`](#type-uint32) `|` `null`
+    * `num_transactions`: [`Uint32`](#type-uint32) `|` `null`
```
Comment (quake) [inline rpc/README.md:1379]: wrong params name?
```suggestion
    * min_transactions`: [`Uint32`](#type-uint32) `|` `null`
```

File: rpc/src/module/chain.rs [hunk 7]
```diff
@@ -1645,4 +1692,124 @@ impl ChainRpc for ChainRpcImpl {
...
+        let tip = snapshot.tip_header().number();
+        let mut match_blocks = Vec::with_capacity(blocks_to_scan as usize);
+        let mut block_counter = 0;
+        let mut tx_counter = 0;
+
+        // get block hash and check if contains enough transaction
+        let mut hash = snapshot.get_block_hash(tip).ok_or_else(|| {
+            RPCError::custom(RPCError::ChainIndexIsInconsistent, "get_block_hash error")
+        })?;
+        while block_counter < blocks_to_scan {
```
Comment (quake) [inline rpc/src/module/chain.rs:1727]: please use block number to simplify the code, pseudocode:

```
for block_number in (tip_number..tip_number - blocks_to_scan) {
    let block_hash = snapshot.get_block_hash(block_number);
    let block_ext = snapshot.get_block_tx_stat(block_hash);
    ...
}
```

**Review by quake — CHANGES_REQUESTED** [review]

File: chain/src/chain.rs [hunk 8]
```diff
@@ -763,13 +763,34 @@ impl ChainService {
...
-                                        cache_entries.into_iter().map(|entry| entry.fee).collect();
+                                        cache_entries.iter().map(|entry| entry.fee).collect();
                                     txn.attach_block(b)?;
                                     attach_block_cell(txn, b)?;
                                     let mut mut_ext = ext.clone();
                                     mut_ext.verified = Some(true);
                                     mut_ext.txs_fees = txs_fees;
                                     txn.insert_block_ext(&b.header().hash(), &mut_ext)?;
+
+                                    // block tx statistics will not include cellbase tx
```
Comment (doitian) [inline chain/src/chain.rs:774]: Please add a feature switch in the config file and disable it by default. Also mention the config option in the RPC document that if users want to use this RPC, they must enable the feature.

File: rpc/README.md (hunk 6 above)
Comment (doitian) [reply rpc/README.md:1379]: Missing a \` in the suggested change.

File: rpc/src/module/chain.rs [hunk 9]
```diff
@@ -1645,4 +1692,124 @@ impl ChainRpc for ChainRpcImpl {
...
+                "chain hasn't enough transactions in {} recent blocks",
+                blocks_to_scan
+            )));
+        }
+
+        let mut fee_vec: Vec<u64> = Vec::with_capacity(blocks_to_scan as usize);
+        for hash in &match_blocks {
+            let stat = snapshot.get_block_tx_stat(hash).ok_or_else(|| {
+                RPCError::custom(
+                    RPCError::ChainIndexIsInconsistent,
```
Comment (doitian) [inline rpc/src/module/chain.rs:1763]: It's OK that database has no the stat for a block. Choose a new RPC error for this.

```
-1200: ChainBlockTxStatsAreNotFound
```

**Review by doitian — CHANGES_REQUESTED** [review]

File: rpc/src/module/chain.rs (hunk 9 above)
Comment (chanhsu001) [reply rpc/src/module/chain.rs:1763]: ok, i will modify code as you and quake suggest, thanks!

File: rpc/README.md (hunk 6 above)
Comment (chanhsu001) [reply rpc/README.md:1379]: i change the param name to:  `blocks_to_scan` and `min_transactions`

File: db-schema/src/lib.rs [hunk 10]
```diff
@@ -3,7 +3,7 @@
 /// Column families alias type
 pub type Col = &'static str;
 /// Total column number
-pub const COLUMNS: u32 = 16;
+pub const COLUMNS: u32 = 17;
```
Comment (doitian) [inline db-schema/src/lib.rs:6]: @zhangsoledad @yangby-cryptape Do we need migration to add a column now?

Comment (yangby-cryptape) [reply db-schema/src/lib.rs:6]: > Do we need migration to add a column now?

Required.

> Or what happens if I open the database with 17 columns in the ckb version which knows only 16 columns?

Panic.

Ref: https://github.com/nervosnetwork/ckb/commit/f782409b7fb22c9d378200540ec5bae8a0f12858

Comment (chanhsu001) [reply db-schema/src/lib.rs:6]: so i need add migration part, right?  and migration still need to done with hardfork? 
please tell me more about migration.

Comment (doitian) [reply db-schema/src/lib.rs:6]: https://github.com/nervosnetwork/ckb/pull/2878

Comment (chanhsu001) [reply db-schema/src/lib.rs:6]: already add  db migration commit

**Comment by chanhsu001** [comment]
close temporarily, re-open if later needed

**Pull request closed without merging**